modified guess_atom_element for more accurate guess

pillose opened the pull request:
Fixes #4167

Changes made in this Pull Request:
 - Suppose the function `guess_atom_element()` receives 'N0A' as an argument 
 - With previous code, 
  `name = 'NA'  # interpreted as sodium`
  Which substitute '0' to ''

 - With new code, 
 `name = ['N', 'A'][0]   #== 'N"`


PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?


<!-- readthedocs-preview mdanalysis start -->
----
:books: Documentation preview :books:: https://mdanalysis--4168.org.readthedocs.build/en/4168/

<!-- readthedocs-preview mdanalysis end -->

pillose committed:
modified guess_atom_element for more accurate guess

hmacdope reviewed (changes requested):
Hi @pillose! Thanks for the contribution!

Would you be able to add a test of the new behaviour? If you need help getting started with the test suite, feel free to ping me.  You will also need to add an entry to `package/CHANGELOG`.

Also as you are a new contributor (welcome!) please add yourself to `package/AUTHORS` and also introduce yourself on the [developer mailing list](https://groups.google.com/g/mdnalysis-devel) or [discord](https://discord.gg/sBXr2eJm).

IAlibay commented:
@pillose this PR has been opened for a little while, are you still interested in contributing this code?

pillose committed:
Merge branch 'MDAnalysis:develop' into develop

Commit:
add test for new behavior

Commit:
Merge branch 'develop' of github.com:pillose/mdanalysis into develop

Commit:
updated AUTHORS and CHANGELOG

pillose commented:
@hmacdope I added some test and change log. Also, I added my name(Geongi Moon) to `package/AUTHORS`.
I apologize for missing your comment.

@IAlibay Thank you for the reminder.

hmacdope commented on package/CHANGELOG:
```diff
@@ -19,6 +19,7 @@ The rules for this file:
 
 Fixes
   * Fix Boltzmann typo in `units.py` (PR #4214, Issue #4213)
+  * Fix Atom type guessing error (PR #4168, Issue #4167)
```
Order this newest first.

hmacdope commented on package/MDAnalysis/topology/guessers.py:
```diff
@@ -204,7 +204,12 @@ def guess_atom_element(atomname):
         # strip symbols and numbers
         no_symbols = re.sub(SYMBOLS, '', atomname)
-        name = re.sub(NUMBERS, '', no_symbols).upper()
+        no_numbers = re.split(NUMBERS, no_symbols)
```
Add some comments as to why this is now in here.

hmacdope reviewed (changes requested):
Looks good, two little comments. Did you end up introducing yourself on mailing list or discord :) ?

IAlibay reviewed (changes requested):
@pillose - We are changing the license under which future contributions are released. Could you please confirm that you agree to releasing this code under the terms of the [LGPLv2.1](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html) and that your contribution also adheres to the [developer certificate of origion](https://developercertificate.org/)?

Commit:
revised the order of CHANGELOG

Commit:
add commnets and modified the code a little

Commit:
Merge branch 'develop' of github.com:pillose/mdanalysis into develop

Commit:
add my username

RMeli commented on package/CHANGELOG:
```diff
@@ -14,11 +14,12 @@ The rules for this file:
 ------------------------------------------------------------------------------
 
-??/??/?? IAlibay, jaclark5, MohitKumar020291, orionarcher, xhgchen, ztimol
+??/??/?? IAlibay, jaclark5, MohitKumar020291, orionarcher, xhgchen, ztimol, pillose
```
```suggestion
??/??/?? IAlibay, jaclark5, MohitKumar020291, orionarcher, xhgchen, ztimol, 
         pillose
```

RMeli commented on package/AUTHORS:
```diff
@@ -219,7 +219,11 @@ Chronological list of authors
   - Daniel J. Evans
   - Mohit Kumar
   - Shubham Kumar
+<<<<<<< HEAD
+  - GeongiMoon
+=======
   - Zaheer Timol
+>>>>>>> 66a2b215760750cdf5e62725249c6971e3ac8b34
```
```suggestion
  - Zaheer Timol
  - GeongiMoon
```

RMeli commented on package/MDAnalysis/topology/guessers.py:
```diff
@@ -202,9 +202,17 @@ def guess_atom_element(atomname):
+        no_numbers = re.split(NUMBERS, no_symbols)
+        no_numbers = list(filter(None, no_numbers)) #remove ''
+        # if no_numbers is not empty, use the first element of no_numbers
+        if no_numbers:
+            name = list(filter(None, no_numbers))[0].upper()
+        else:
+            name = ''
```
```suggestion
        name = list(filter(None, no_numbers))[0].upper() if no_numbers else ''
```

RMeli reviewed (commented):
Thanks @pillose. Please see https://github.com/MDAnalysis/mdanalysis/pull/4168#pullrequestreview-1554498651, thanks.

pillose commented:
@IAlibay I agree to changing the license.
@hmacdope I just introduced myself on mailing list.
@RMeli Thank you for the reminder.

pillose committed:
Update package/CHANGELOG

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

pillose committed:
Update package/AUTHORS

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

pillose committed:
Update package/MDAnalysis/topology/guessers.py

Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

Commit:
Merge branch 'develop' of github.com:pillose/mdanalysis into develop

pillose committed:
Merge branch 'develop' into develop

hmacdope commented on package/AUTHORS:
```diff
@@ -220,6 +220,7 @@ Chronological list of authors
   - Mohit Kumar
   - Shubham Kumar
   - Zaheer Timol
+  - GeongiMoon
```
Needs a space, I am assuming?

hmacdope replied:
If not let me know, other than that looks good.

pillose replied:
I missed a space... Thank you!

pillose committed:
Update AUTHORS

hmacdope commented on testsuite/MDAnalysisTests/topology/test_guessers.py:
```diff
@@ -104,6 +104,9 @@ def test_guess_atom_element_1H(self):
         ('zn', 'ZN'),
         ('Ca2+', 'CA'),
         ('CA', 'C'),
+        ('N0A', 'N'),
```
One more nit sorry could you test  Na+ -> Na?

pillose replied:
I just added two tests for Na+ and Cu2+

pillose committed:
add guesser tests

pillose committed:
Update test_guessers.py

IAlibay commented on package/MDAnalysis/topology/guessers.py:
```diff
@@ -202,9 +202,14 @@ def guess_atom_element(atomname):
+        no_numbers = re.split(NUMBERS, no_symbols)
+        no_numbers = list(filter(None, no_numbers)) #remove ''
+        # if no_numbers is not empty, use the first element of no_numbers
+        name = list(filter(None, no_numbers))[0].upper() if no_numbers else ''
```
Apologies if I'm missing something obvious, but do you need to apply filter again here? You've already removed the `''` entries and you shouldn't have any `None` entries ?

pillose replied:
Sorry, I think I made a mistake while modifying the code...

IAlibay reviewed (changes requested):
The one question from me.

pillose committed:
Update guessers.py

remove unnecessary code

IAlibay reviewed (approved):
Thanks, sorry for this taking so long. lgtm!

hmacdope reviewed (approved):
LGTM also thanks @pillose

hmacdope commented:
Cycling CI

pillose committed:
Merge branch 'develop' into develop
